cpp/lib/src/LocalDataDB.cc

```diff
@@ -0,0 +1,215 @@
+
+void LocalDataDB::insertOrReplace(const std::string &title_ppn, const std::vector<std::string> &local_fields) {
+    // 1. Clear out any local PPNs associated with "title_ppn":
+    db_connection_->queryOrDie("SELECT local_fields FROM local_data WHERE title_ppn = "
+                               + db_connection_->escapeAndQuoteString(title_ppn));
+    auto result_set(db_connection_->getLastResultSet());
+    if (not result_set.empty()) {
+        const auto row(result_set.getNextRow());
+        const auto previous_local_fields(BlobToLocalFieldsVector(row["local_fields"], title_ppn));
+        const auto local_ppns(ExtractLocalPPNsFromLocalFieldsVector(previous_local_fields));
+        for (const auto &local_ppn : local_ppns)
+            db_connection_->queryOrDie("DELETE * FROM local_ppns_to_title_ppns_map WHERE local_ppn="
+                                       + db_connection_->escapeAndQuoteString(local_ppn));
+    }
+
+    // 2. Replace or insert the local data keyed by the title PPN's:
+    db_connection_->queryOrDie("REPLACE INTO local_data (title_ppn, local_fields) VALUES("
+                               + db_connection_->escapeAndQuoteString(title_ppn) + ","
+                               + db_connection_->sqliteEscapeBlobData(ConvertLocalFieldsVectorToBlob(local_fields)) + ")");
+
+    // 3. Insert the mappings from the local PPN's to the title PPN's:
+    for (const auto &local_ppn : ExtractLocalPPNsFromLocalFieldsVector(local_fields)) {
+        db_connection_->queryOrDie("INSERT INTO local_ppns_to_title_ppns_map (local_ppn, title_ppn) VALUES("
+                                   + db_connection_->escapeAndQuoteString(local_ppn) + ","
+                                   + db_connection_->escapeAndQuoteString(title_ppn) + ")");
+    }
+}
+
+
+std::vector<std::string> LocalDataDB::getLocalFields(const std::string &title_ppn) const {
+    db_connection_->queryOrDie("SELECT local_fields FROM local_data WHERE title_ppn = "
+                               + db_connection_->escapeAndQuoteString(title_ppn));
+    auto result_set(db_connection_->getLastResultSet());
+    if (result_set.empty())
+        return {}; // empty vector
+
+    const auto row(result_set.getNextRow());
+    return BlobToLocalFieldsVector(row["local_fields"], title_ppn);
+}
+
+
+bool LocalDataDB::removeTitleDataSet(const std::string &title_ppn) {
+    // 1. Clear out any local PPNs associated with "title_ppn":
+    db_connection_->queryOrDie("SELECT local_fields FROM local_data WHERE title_ppn = "
+                               + db_connection_->escapeAndQuoteString(title_ppn));
+    auto result_set(db_connection_->getLastResultSet());
+    if (result_set.empty())
+        return false;
+
+    const auto row(result_set.getNextRow());
+    const auto previous_local_fields(BlobToLocalFieldsVector(row["local_fields"], title_ppn));
+    const auto local_ppns(ExtractLocalPPNsFromLocalFieldsVector(previous_local_fields));
+    for (const auto &local_ppn : local_ppns)
```

Comment:
Couldn't this be solved via a foreign key with cascade delete?

Reply:
I don't think so because the local PPN's are not column values in the other table.